R/JoinKeys.R

```diff
@@ -164,12 +164,84 @@ JoinKeys <- R6::R6Class( # nolint
       } else {
         cat("An empty JoinKeys object.")
       }
+      invisible(self)
+    },
+    #' @description
+    #' Sets the parents of the datasets.
+    #'
+    #' @param named_list Named (`list`) of the parents datasets.
+    #'
+    #' @return (`self`) invisibly for chaining
+    set_parents = function(named_list) {
+      for (dataset in names(named_list)) {
+        if (is.null(self$get_parent(dataset))) {
+          private$parents[[dataset]] <- named_list[[dataset]]
+        }
+      }
+      invisible(self)
+    },
+    #' @description
+    #' Gets the parent of the desired dataset.
+    #'
+    #' @param dataname (`character`) name of the dataset.
+    #' @return (`character`) the parent of the desired dataset
+    get_parent = function(dataname) {
+      private$parents[[dataname]]
+    },
+    #' @description
+    #' Gets the parents of the datasets.
+    #'
+    #' @return (`list`) A named list of the parents of all datasets
+    get_parents = function() {
+      private$parents
+    },
+    #' @description
+    #' Updates the keys of the datasets based on the parents.
+    #'
+    #' @return (`self`) invisibly for chaining
+    update_keys_given_parents = function() {
+      datanames <- names(self$get())
+      duplicate_pairs <- list()
+      for (d1 in datanames) {
+        d1_pk <- self$get(d1, d1)
+        d1_parent <- self$get_parent(d1)
+        for (d2 in datanames) {
+          if (paste(d2, d1) %in% duplicate_pairs) {
+            next
+          }
+          if (length(self$get(d1, d2)) == 0) {
+            d2_parent <- self$get_parent(d2)
+            d2_pk <- self$get(d2, d2)
+
+            fk <- if (identical(d1, d2_parent)) {
+              # first is parent of second -> parent keys -> first keys
+              d1_pk
+            } else if (identical(d1_parent, d2)) {
+              # second is parent of first -> parent keys -> second keys
+              d2_pk
+            } else if (identical(d1_parent, d2_parent) && length(d1_parent) > 0) {
+              # both has the same parent -> parent keys
+              self$get(d1_parent, d1_parent)
+            } else {
+              # cant find connection - leave empty
+              next
+            }
+            self$mutate(d1, d2, fk)
+            duplicate_pairs <- append(duplicate_pairs, paste(d1, d2))
+          }
+        }
+      }
+      # check parent child relation
+      private$check_parent_child()
+
       invisible(self)
     }
   ),
   ## __Private Fields ====
   private = list(
     .keys = list(),
+    primary_keys = list(),
```

Comment:
I don't think we need this as this can be stored in .keys as well (.keys[d, d]) and we can have a helper function `get_primary_keys(d)`